src/proto/streams/flow_control.rs

```diff
@@ -133,8 +142,14 @@ impl FlowControl {
             self.available
         );
         // This should not be able to overflow `window_size` from the bottom.
-        self.window_size -= sz as i32;
-        self.available = self.available.saturating_sub(sz);
+        self.window_size -= sz;
+        self.available -= sz;
+        // This should only be called on Send flow controllers. In this case,
+        // available capacity must have been reserved, and can be freed. We
+        // don't need to go negative here.
+        if self.available.0 < 0 {
```

Comment:
I'm not sure that I follow this scenario & comment. If this is only called on the `Send` flow controller (aka, *we* are sending out `DATA` frames to a remote) then how does this happen? It should be impossible to go below zero available? (I'm probably forgetting the exact scenario).

Reply:
I was confused by this too, but the tests found it. The flow controls on the send side of a stream don't gain available capacity until it's been reserved specifically. In fact, this kind of confuses me. It seems that when a send side reserves capacity, the number gets added to `available`, almost as if this means "bytes we have reserved locally but could give back to the connection".

The situation that broke without this condition is reserving 11 bytes, so then `available == 11`, and then getting a SETTINGS frame reducing initial window size to 0, so it was called as `dec_window(65_535)`, resulting in a negative "available".

Reply:
To try to make this less confusing, I've completely removed touching the `available` capacity from `dec_window`. It is now two separate actions. In `Send`, after decreasing the window, the code then takes away `cmp::min(dec, available)` capacity as well.